File: legacy/nest-server/modules/performance/performance.controller.ts
import {
  Controller,
  Get,
  Post,
  Patch,
  Param,
  Query,
  Body,
  Req,
  UseGuards,
} from '@nestjs/common';
import { PerformanceService } from './performance.service';
import type { Request } from 'express';
import type {
  ReviewItem,
  GoalSettingItem,
  ApproveGoalRequest,
  FinalReviewRequest,
  CreatePerformanceRequest,
  SelectTemplateRequest,
  CalibrationReviewRequest,
} from '@shared/api.interface';
import { SessionUserGuard } from '../../guards/session-user.guard';

@Controller('api/performances')
@UseGuards(SessionUserGuard)
export class PerformanceController {
  constructor(private readonly performanceService: PerformanceService) {}

  @Get()
  async list(
    @Req() req: Request,
    @Query('status') status?: string,
    @Query('focus') focus?: string,
    @Query('period') period?: string,
    @Query('departmentId') departmentId?: string,
    @Query('employeeName') employeeName?: string,
    @Query('page') page?: string,
    @Query('pageSize') pageSize?: string,
  ) {
    const userId = req.userContext!.userId!;
    const userRole = await this.performanceService.getUserRole(userId);
    const result = await this.performanceService.list(userId, {
      status,
      focus,
      period,
      departmentId,
      employeeName,
      page: page ? parseInt(page, 10) : 1,
      pageSize: pageSize ? parseInt(pageSize, 10) : 20,
    });
    return { ...result, userRole };
  }

  @Get('create/month-periods')
  async listMonthPeriodsForCreate(@Req() req: Request) {
    return this.performanceService.listMonthPeriodsForCreate(req.userContext!.userId!);
  }

  @Get('export')
  async export(
    @Req() req: Request,
    @Query('status') status?: string,
    @Query('focus') focus?: string,
    @Query('period') period?: string,
    @Query('departmentId') departmentId?: string,
    @Query('employeeName') employeeName?: string,
  ) {
    const userId = req.userContext!.userId!;
    return this.performanceService.exportData(userId, {
      status,
      focus,
      period,
      departmentId,
      employeeName,
    });
  }

  @Get('calibration/supervisor-queue')
  async listSupervisorCalibrationQueue(
    @Req() req: Request,
    @Query('period') period?: string,
    @Query('departmentId') departmentId?: string,
    @Query('employeeName') employeeName?: string,
    @Query('page') page?: string,
    @Query('pageSize') pageSize?: string,
  ) {
    const userId = req.userContext!.userId!;
    return this.performanceService.listSupervisorCalibrationQueue(userId, {
      period,
      departmentId,
      employeeName,
      page: page ? parseInt(page, 10) : 1,
      pageSize: pageSize ? parseInt(pageSize, 10) : 20,
    });
  }

  @Get(':id')
  async getDetail(@Req() req: Request, @Param('id') id: string) {
    const userId = req.userContext!.userId!;
    return this.performanceService.getDetail(userId, id);
  }

  @Patch(':id')
  async saveDraft(
    @Req() req: Request,
    @Param('id') id: string,
    @Body()
    body: {
      reviewType: 'goal' | 'self' | 'manager' | 'dotted_manager';
      content: ReviewItem[] | GoalSettingItem[];
      personalSummary?: string;
    },
  ) {
    return this.performanceService.saveDraft(req.userContext!.userId!, id, body);
  }

  @Post(':id/submit')
  async submit(
    @Req() req: Request,
    @Param('id') id: string,
    @Body()
    body: {
      reviewType: 'goal' | 'self' | 'manager' | 'dotted_manager';
      content: ReviewItem[] | GoalSettingItem[];
      personalSummary?: string;
    },
  ) {
    return this.performanceService.submit(req.userContext!.userId!, id, body);
  }

  @Post(':id/reject')
  async reject(
    @Req() req: Request,
    @Param('id') id: string,
    @Body() body: { reason: string },
  ) {
    return this.performanceService.reject(req.userContext!.userId!, id, body.reason);
  }

  @Post(':id/approve-goal')
  async approveGoal(
    @Req() req: Request,
    @Param('id') id: string,
    @Body() body: ApproveGoalRequest,
  ) {
    return this.performanceService.approveGoal(req.userContext!.userId!, id, body);
  }

  @Post(':id/final-review')
  async finalReview(
    @Req() req: Request,
    @Param('id') id: string,
    @Body() body: FinalReviewRequest,
  ) {
    return this.performanceService.finalReview(req.userContext!.userId!, id, body);
  }

  @Post()
  async create(
    @Req() req: Request,
    @Body() body: CreatePerformanceRequest,
  ) {
    return this.performanceService.createBatch(req.userContext!.userId!, body);
  }

  @Post(':id/select-template')
  async selectTemplate(
    @Req() req: Request,
    @Param('id') id: string,
    @Body() body: SelectTemplateRequest,
  ) {
    return this.performanceService.selectTemplate(
      req.userContext!.userId!,
      id,
      body.templateId,
    );
  }

  @Post(':id/calibrate')
  async calibrate(
    @Req() req: Request,
    @Param('id') id: string,
    @Body() body: CalibrationReviewRequest,
  ) {
    return this.performanceService.calibrate(
      req.userContext!.userId!,
      id,
      body,
    );
  }
}

File: server-java/src/main/java/com/jixiao2/server/performance/PerformanceController.java
package com.jixiao2.server.performance;

import com.jixiao2.server.web.CurrentUser;
import java.util.Map;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/performances")
public class PerformanceController {

  private final PerformanceService performanceService;

  public PerformanceController(PerformanceService performanceService) {
    this.performanceService = performanceService;
  }

  @GetMapping
  public Map<String, Object> list(
      @CurrentUser String userId,
      @RequestParam(required = false) String status,
      @RequestParam(required = false) String focus,
      @RequestParam(required = false) String period,
      @RequestParam(required = false) String subjectCode,
      @RequestParam(required = false) String departmentId,
      @RequestParam(required = false) String employeeName,
      @RequestParam(required = false, defaultValue = "1") int page,
      @RequestParam(required = false, defaultValue = "20") int pageSize) {
    Map<String, Object> result =
        performanceService.list(
            userId, status, focus, period, subjectCode, departmentId, employeeName, page, pageSize);
    result.put("userRole", performanceService.getUserRole(userId));
    return result;
  }

  @GetMapping("/create/month-periods")
  public Map<String, Object> listMonthPeriodsForCreate(@CurrentUser String userId) {
    return performanceService.listMonthPeriodsForCreate(userId);
  }

  @GetMapping("/filter/month-periods")
  public Map<String, Object> listMonthPeriodsForFilter(@CurrentUser String userId) {
    return performanceService.listMonthPeriodsForFilter(userId);
  }

  @GetMapping("/export")
  public Map<String, Object> export(
      @CurrentUser String userId,
      @RequestParam(required = false) String status,
      @RequestParam(required = false) String focus,
      @RequestParam(required = false) String period,
      @RequestParam(required = false) String subjectCode,
      @RequestParam(required = false) String departmentId,
      @RequestParam(required = false) String employeeName) {
    return performanceService.exportData(
        userId, status, focus, period, subjectCode, departmentId, employeeName);
  }

  @GetMapping("/calibration/supervisor-queue")
  public Map<String, Object> listSupervisorCalibrationQueue(
      @CurrentUser String userId,
      @RequestParam(required = false) String period,
      @RequestParam(required = false) String subjectCode,
      @RequestParam(required = false) String departmentId,
      @RequestParam(required = false) String employeeName,
      @RequestParam(required = false, defaultValue = "1") int page,
      @RequestParam(required = false, defaultValue = "20") int pageSize) {
    return performanceService.listSupervisorCalibrationQueue(
        userId, period, subjectCode, departmentId, employeeName, page, pageSize);
  }

  @GetMapping("/{id}")
  public Map<String, Object> getDetail(@CurrentUser String userId, @PathVariable String id) {
    return performanceService.getDetail(userId, id);
  }

  @DeleteMapping("/{id}")
  public Map<String, Object> delete(@CurrentUser String userId, @PathVariable String id) {
    return performanceService.softDelete(userId, id);
  }

  @PatchMapping("/{id}")
  public Map<String, Object> saveDraft(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    return performanceService.saveDraft(userId, id, body);
  }

  @PostMapping("/{id}/submit")
  public Map<String, Object> submit(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body)
      throws Exception {
    return performanceService.submit(userId, id, body);
  }

  @PostMapping("/{id}/reject")
  public Map<String, Object> reject(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    String reason = body.get("reason") == null ? "" : String.valueOf(body.get("reason"));
    return performanceService.reject(userId, id, reason);
  }

  @PostMapping("/{id}/approve-goal")
  public Map<String, Object> approveGoal(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    return performanceService.approveGoal(userId, id, body);
  }

  @PostMapping("/{id}/final-review")
  public Map<String, Object> finalReview(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    return performanceService.finalReview(userId, id, body);
  }

  @PostMapping
  public Map<String, Object> create(
      @CurrentUser String userId, @RequestBody Map<String, Object> body) {
    return performanceService.createBatch(userId, body);
  }

  @PostMapping("/{id}/select-template")
  public Map<String, Object> selectTemplate(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    String templateId = body.get("templateId") == null ? "" : String.valueOf(body.get("templateId"));
    return performanceService.selectTemplate(userId, id, templateId);
  }

  @PutMapping("/{id}/goal-indicators")
  public Map<String, Object> saveGoalIndicators(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    return performanceService.saveGoalIndicators(userId, id, body);
  }

  @PostMapping("/{id}/calibrate")
  public Map<String, Object> calibrate(
      @CurrentUser String userId, @PathVariable String id, @RequestBody Map<String, Object> body) {
    return performanceService.calibrate(userId, id, body);
  }

  @PostMapping("/{id}/confirm-result")
  public Map<String, Object> confirmResult(
      @CurrentUser String userId, @PathVariable String id) {
    return performanceService.confirmResult(userId, id);
  }
}
